File: BackEnd_Learn/Java/auth/src/main/java/io/github/leeseojune53/auth/security/jwt/JwtTokenProvider.java
package io.github.leeseojune53.auth.security.jwt;

import io.github.leeseojune53.auth.security.auth.AuthDetailsService;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import java.util.Base64;
import java.util.Date;

@Component
@RequiredArgsConstructor
public class JwtTokenProvider {

    @Value("${jwt.secret}")
    private String secretKey;

    private final AuthDetailsService authDetailsService;

    public String generateAccessToken(String id) {
        return Jwts.builder()
                .signWith(SignatureAlgorithm.HS256, getSigningKey())
                .setHeaderParam("typ", "JWT")
                .claim("type", "access")
                .setSubject(id)
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + 6000 * 1000))
                .compact();
    }

    public String resolveToken(HttpServletRequest httpServletRequest) {
        String bearer = httpServletRequest.getHeader("Authorization");
        if(bearer != null && bearer.startsWith("Bearer"))
            return bearer.substring(7);
        return null;
    }

    public boolean validateToken(String token) {
        return getTokenBody(token).getExpiration()
                .after(new Date());
    }

    public Authentication authentication(String token) {
        UserDetails userDetails = authDetailsService.loadUserByUsername(getId(token));
        return new UsernamePasswordAuthenticationToken(userDetails, "", userDetails.getAuthorities());
    }

    private Claims getTokenBody(String token) {
        return Jwts.parser().setSigningKey(getSigningKey())
                .parseClaimsJws(token).getBody();
    }

    private String getId(String token) {
        return getTokenBody(token).getSubject();
    }

    private String getSigningKey() {
        return Base64.getEncoder().encodeToString(secretKey.getBytes());
    }

}

File: BackEnd_Learn/kotlin/auth/src/main/kotlin/io/github/leeseojune53/auth/security/jwt/JwtTokenProvider.kt
package io.github.leeseojune53.auth.security.jwt

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*
import javax.servlet.http.HttpServletRequest

@Component
class JwtTokenProvider {

    @Value("jwt.secret")
    private lateinit var secretKey: String

    @Value("jwt.access.exp")
    private var accessExp: Long = 0

    fun generateAccessToken(id: String) : String {
        return Jwts.builder()
            .signWith(SignatureAlgorithm.HS256, secretKey)
            .setHeaderParam("typ", "JWT")
            .claim("type", "access")
            .setSubject(id)
            .setIssuedAt(Date())
            .setExpiration(Date(System.currentTimeMillis() + 6000 * 1000))
            .compact()
    }

    fun resolveToken(httpServletRequest: HttpServletRequest): String {

        var bearer = httpServletRequest.getHeader("Authorization");
        if(bearer != null && bearer.toString())
    }

}